pkg/storage/unified/search/bleve.go

```diff
@@ -554,6 +550,60 @@ func toBleveSearchRequest(req *resource.ResourceSearchRequest, access authz.Acce
 		searchrequest.Query = bleve.NewConjunctionQuery(queries...) // AND
 	}
 
+	// Can we remove this? Is access ever nil?
+	if access != nil {
+		auth, ok := claims.From(ctx)
+		if !ok {
+			return nil, resource.AsErrorResult(fmt.Errorf("missing claims"))
+		}
+		folderChecker, err := access.Compile(ctx, auth, authz.ListRequest{
+			Namespace: b.key.Namespace,
+			Group:     b.key.Group,
+			Resource:  "folder",
+			Verb:      utils.VerbList,
+		})
+		if err != nil {
+			return nil, resource.AsErrorResult(err)
+		}
+		dashboardChecker, err := access.Compile(ctx, auth, authz.ListRequest{
+			Namespace: b.key.Namespace,
+			Group:     b.key.Group,
+			Resource:  "dashboard",
+			Verb:      utils.VerbList,
+		})
+		if err != nil {
+			return nil, resource.AsErrorResult(err)
+		}
+
+		permissionsQuery := &permissionScopedQuery{
+			Query: searchrequest.Query,
+			permissionsFilter: func(d *search.DocumentMatch) bool {
+				// The internal ID has the format: <namespace>/<group>/<resourceType>/<name>
+				// Only the internal ID is present on the document match here
+				id := string(d.IndexInternalID)
+				parts := strings.Split(id, "/")
+				// Exclude doc if id isn't expected format
+				if len(parts) != 4 {
+					return false
+				}
+				name := parts[len(parts)-1]
+				resourceType := parts[2]
+				if resourceType == "folders" {
+					if !folderChecker(b.key.Namespace, name, id) {
+						return false
+					}
+				}
+				if resourceType == "dashboards" {
+					if !dashboardChecker(b.key.Namespace, name, id) {
+						return false
+					}
+				}
+				return true
```

Comment:
What about this implementation? Would it read cleaner?

```suggestion
                if resourceType == "folders" {
					return folderChecker(b.key.Namespace, name, id)
				}
				if resourceType == "dashboards" {
					return dashboardChecker(b.key.Namespace, name, id) {
				}
				return true
```